File: backend/app/schemas/vehicle.py
import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel


class VehicleCreate(BaseModel):
    plate_number: str
    make: str | None = None
    model: str | None = None
    year: int | None = None
    color: str | None = None
    vin: str | None = None
    vehicle_type: str = "motorcycle"
    ownership: str = "company"
    rental_company: str | None = None
    current_mileage: int | None = None
    fuel_type: str | None = None
    assigned_driver_id: uuid.UUID | None = None
    insurance_expiry: date | None = None
    registration_expiry: date | None = None
    notes: str | None = None


class VehicleUpdate(BaseModel):
    plate_number: str | None = None
    make: str | None = None
    model: str | None = None
    year: int | None = None
    color: str | None = None
    vehicle_type: str | None = None
    ownership: str | None = None
    rental_company: str | None = None
    current_mileage: int | None = None
    fuel_type: str | None = None
    status: str | None = None
    assigned_driver_id: uuid.UUID | None = None
    insurance_expiry: date | None = None
    registration_expiry: date | None = None
    notes: str | None = None


class VehicleResponse(BaseModel):
    id: uuid.UUID
    tenant_id: uuid.UUID
    plate_number: str
    make: str | None
    model: str | None
    year: int | None
    color: str | None
    vin: str | None
    vehicle_type: str
    ownership: str
    rental_company: str | None
    current_mileage: int | None
    fuel_type: str | None
    status: str
    assigned_driver_id: uuid.UUID | None
    insurance_expiry: date | None
    registration_expiry: date | None
    notes: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class MaintenanceCreate(BaseModel):
    vehicle_id: uuid.UUID
    driver_id: uuid.UUID | None = None
    date: date
    category: str
    type: str
    description: str | None = None
    cost: Decimal | None = None
    vendor: str | None = None
    mileage_at_service: int | None = None
    notes: str | None = None


class MaintenanceUpdate(BaseModel):
    status: str | None = None
    cost: Decimal | None = None
    vendor: str | None = None
    spare_vehicle_id: uuid.UUID | None = None
    mechanic_dispatched: bool | None = None
    notes: str | None = None


class MaintenanceResponse(BaseModel):
    id: uuid.UUID
    tenant_id: uuid.UUID
    vehicle_id: uuid.UUID
    driver_id: uuid.UUID | None
    date: date
    category: str
    type: str
    description: str | None
    cost: Decimal | None
    vendor: str | None
    mileage_at_service: int | None
    duration_hours: Decimal | None
    receipt_url: str | None
    status: str
    source: str
    spare_vehicle_id: uuid.UUID | None
    mechanic_dispatched: bool | None
    police_report_url: str | None
    medical_report_url: str | None
    notes: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class InspectionCreate(BaseModel):
    vehicle_id: uuid.UUID
    driver_id: uuid.UUID
    shift_id: uuid.UUID | None = None
    checklist: dict = {}
    photos: list = []
    overall_status: str = "pass"
    notes: str | None = None
    location_lat: float | None = None
    location_lng: float | None = None


class InspectionResponse(BaseModel):
    id: uuid.UUID
    tenant_id: uuid.UUID
    vehicle_id: uuid.UUID
    driver_id: uuid.UUID
    shift_id: uuid.UUID | None
    inspected_at: datetime
    checklist: dict
    photos: list
    overall_status: str
    notes: str | None
    location_lat: float | None
    location_lng: float | None

    model_config = {"from_attributes": True}

File: frontend/src/types/vehicle.ts
export interface Vehicle {
  id: string;
  tenant_id: string;
  plate_number: string;
  make: string | null;
  model: string | null;
  year: number | null;
  color: string | null;
  vin: string | null;
  vehicle_type: string;
  ownership: string;
  rental_company: string | null;
  current_mileage: number | null;
  fuel_type: string | null;
  status: string;
  assigned_driver_id: string | null;
  insurance_expiry: string | null;
  registration_expiry: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface VehicleCreate {
  plate_number: string;
  make?: string;
  model?: string;
  year?: number;
  color?: string;
  vin?: string;
  vehicle_type?: string;
  ownership?: string;
  rental_company?: string;
  current_mileage?: number;
  fuel_type?: string;
  assigned_driver_id?: string;
  insurance_expiry?: string;
  registration_expiry?: string;
  notes?: string;
}

export interface VehicleUpdate extends Partial<VehicleCreate> {
  status?: string;
}

export interface VehicleFilters {
  status?: string;
  vehicle_type?: string;
  ownership?: string;
  search?: string;
  page?: number;
  per_page?: number;
}

export interface MaintenanceRecord {
  id: string;
  tenant_id: string;
  vehicle_id: string;
  driver_id: string | null;
  date: string;
  category: string;
  type: string;
  description: string | null;
  cost: number | null;
  vendor: string | null;
  mileage_at_service: number | null;
  duration_hours: number | null;
  receipt_url: string | null;
  status: string;
  source: string;
  spare_vehicle_id: string | null;
  mechanic_dispatched: boolean | null;
  police_report_url: string | null;
  medical_report_url: string | null;
  notes: string | null;
  created_at: string;
}

export interface VehicleInspection {
  id: string;
  tenant_id: string;
  vehicle_id: string;
  driver_id: string;
  shift_id: string | null;
  inspected_at: string;
  checklist: Record<string, unknown>;
  photos: string[];
  overall_status: string;
  notes: string | null;
  location_lat: number | null;
  location_lng: number | null;
}
